### ISSUE TITLE
XDMA Advanced topics

### ISSUE THREAD

**Issue by AGenchev** [mwrnd/notes#1]
Hello! I'm glad to see more interesting experiments!
I saw you posted a link to [Xilinx DMA PCIe Tutorial](https://www.linkedin.com/pulse/xilinx-dma-pcie-tutorial-part-1-roy-messinger) by Roy Messinger. I read it. He implements XDMA with descriptor bypass, which is interesting but not became clear what exactly happens. 
Do you plan to test the descriptor bypass feature ? It seems supported by the driver w/o modification.
AFAIK it is convinent when the memory is not contiguous. 

Also, I try to imagine how they do double buffering in stream mode - there should be a buffer on the host and the user/driver should get notification/IRQ when a buffer is filled so to replace it with another while the DMA is writing into its current buffer.

**Comment by mwrnd** [comment]
Yes, my understanding of the use case for [descriptor](https://docs.xilinx.com/r/en-US/pg195-pcie-dma/Descriptors) bypass is to sequence non-contiguous memory accesses ([scatter-gather](https://en.wikipedia.org/wiki/Gather/scatter_(vector_addressing))). I do not see how it would improve large contiguous transfers or AXI-Stream when using XDMA.

The [Xilinx DMA PCIe Tutorial](https://www.linkedin.com/pulse/xilinx-dma-pcie-tutorial-part-1-roy-messinger) may be a bit too out of date now. With the current state of the XDMA driver there is no need to organize DMA transfers.

> c2h double buffering in stream mode

Don't know. I would focus on [FIFOs](https://github.com/mwrnd/innova2_experiments/tree/main/xdma_stream_512bit#fifos) and the [`_tready` signal](https://zipcpu.com/blog/2022/02/23/axis-abort.html)([1](https://support.xilinx.com/s/question/0D52E00006hpQAqSAM/instantaneous-asynchronous-back-pressure-on-block-achievable-through-axi4-stream-protocol?language=en_US)). Best performance comes from having large enough FIFOs to allow data to flow from/into your application as soon as it is ready.

![XDMA Stream Loopback](https://github.com/mwrnd/notes/assets/38073881/18920210-9cec-4cf7-809c-d9fa6475bcb4)

**Comment by AGenchev** [comment]
What I was thinking is that (still haven't read the xdma driver source) the XDMA driver setups each transfer (maps buffers?) before submit it to the BM DMA of the IP Core in the FPGA, which leads to DMA engine idling. So they provided 4 channels to somehow allow an user app to counter that by submitting overlapped requests to the 4 channels. How I imagine descriptor bypass should work:
- somebody (user app, driver) allocates nc buffers and creates SG descriptor tables for them also in nc memory.
- driver setups the address of the first SG descriptor. The last descriptor closes the ring, pointing to the first descriptor as next_descriptor. The descriptors have flags, which the IP core checks to know whether to stop, it also (optionally) marks which descriptor has been used (written to the memory block).
- when the start is issued, the IP core reads the first descriptor, initiates DMA over PCIe to it, and initiates prefetch of N next descriptors in cache BRAM so the next TLP to have readily available PHY memory address.
- somehow the driver should get notification to process an available block of data - either by MSI, generated by some criteria (threshold of written memory blocks). It can intervene and switch the next_descriptor of the last descriptor for example to point to a new group of descriptors, so to enable the currently available block of data to be processed without worry of being overwritten in the meantime and to allow memory copy to be avoided.
I'll dig into the driver once i manage to start the test pattern generator and video write buffer to work (shame on me).

**Comment by mwrnd** [comment]
Yes, for XDMA MM (Memory-Mapped) the 4 channels are interchangeable and each [corresponds to a thread](https://github.com/Xilinx/dma_ip_drivers/blob/a93d4a4870e41d152b33aebb3f869eefb11aa691/XDMA/linux-kernel/xdma/libxdma.c#L107) so you can overlap read/write requests and let the driver sort things out.

For XDMA Stream, each C2H/H2C channel corresponds to its own circuit, `S_AXIS_C2H_?`/`M_AXIS_H2C_?`. If you are working on Video I recommend you look into the [AXI4-Stream Video Tutorials](https://support.xilinx.com/s/article/Video-Series-1-Introduction-to-Digital-Video?language=en_US), [pg231](https://docs.xilinx.com/r/en-US/pg231-v-proc-ss/AXI4-Stream-Video), and [ug934](https://docs.xilinx.com/r/en-US/ug934_axi_videoIP).

[XDMA Descriptor Bypass mentions there is an 8-deep queue for descriptors](https://docs.xilinx.com/r/en-US/pg195-pcie-dma/Descriptor-Bypass?tocId=3aLBs6k9P1HtbELPvnwEXQ).

I stumbled across the [Descriptor Bypass Controller for Xilinx XDMA](https://github.com/rsarwar87/xdma_dsc_byp_cltr) project that I will look through.

**Comment by Prandr** [comment]
Since we are at discussion of advanced topics, what the DMA bypass (not to be confused with _descriptor_ bypass) is good for?

**Comment by mwrnd** [comment]
My best guess is to be able to implement your own DMA engine. If your project has defined and specific access sequences you should be able to optimize for it.

**Comment by Prandr** [comment]
> My best guess is to be able to implement your own DMA engine. If your project has defined and specific access sequences you should be able to optimize for it.

That is what _descriptor_ bypass is for.
DMA bypass seems to acts like integrated  AXI-bridge, similar to standalone option for US+. Probably for small size low latency access.

**Comment by Prandr** [comment]
@AGenchev @mwrnd 
In past 12 months I have gone the path from having little idea to becoming an expert  in XDMA and can now fully join the discussion.

I've read Roy's guide and there are points which hint to lack of deep understanding as well as clear errors.

Like mwrnd, I don't see an advantage of using descriptor bypass, except for exotic cases, that would likely require a custom driver. There is an additional complication: DMA addresses (in the descriptors) lie in a special bus address space, that are produced by `dma_map_..` kernel API functions. How the FPGA logic is supposed to get these addresses? I didn't understand that from Roy's guide.

> What I was thinking is that (still haven't read the xdma driver source) the XDMA driver setups each transfer (maps buffers?) before submit it to the BM DMA of the IP Core in the FPGA, which leads to DMA engine idling. So they provided 4 channels to somehow allow an user app to counter that by submitting overlapped requests to the 4 channels. How I imagine descriptor bypass should work:
>....

what you describe is close to how Eli Billauer's [xillybus](http://xillybus.com) is working. It uses arrays of kernel buffers to maintain as continuous data flow as possible. There is a drawback to this method: the data needs to be copied from/to kernel buffers.
I considered to implement [my reworked version of the XDMA driver](https://github.com/Prandr/dma_ip_drivers) like that, but rejected the idea, because XDMA IP lacks necessary flow control facilities. If for example, C2H buffers would fill up and get under threat of overflow, there is no way to pause the transfer.

On the other hand XDMA's method (mapping user buffers into DMA/bus address space) has an advantage, that the data is sent/received directly without need of any additional copying.



> I took a quick look at the BYP controller. It is not extensively documented as code or diagram of working so it needs effort to be understood. The author says "For C2H transfer, allocated a continuous memory", which is strange - the SG DMA engine shouldn't need contiguous memory, it should be able to work with kernel VMA allocated memory or mmapped user process memory. In the common case, descriptors can point to variable sized blocks of physical memory.

I  didn't study it, but what is likely meant that each individual buffer must be continuous to be mapped to DMA address
> I think that if the blocks are allocated with equal size == page size (4k), then working with the DMA engine becomes easier: just using the number of processed descriptors, one can calculate how much memory has been written to. .

That would make very inefficient use of descriptors which support up to 256 MB ( and in fact the original XDMA driver does just like this). Calculate many descriptors you would need for, say, 32 MB transfer.



> > My best guess is to be able to implement your own DMA engine. If your project has defined and specific access sequences you should be able to optimize for it.
> 
> That is what _descriptor_ bypass is for. DMA bypass seems to acts like integrated AXI-bridge, similar to standalone option for US+. Probably for small size low latency access.

I can now answer my own question. I can confirm that I guessed correctly. However, its implementation is broken in AMD's driver. I reworked it in my driver as well.
